controllers/volumereplicationgroup_controller.go

```diff
@@ -541,15 +545,25 @@ func (v *VRGInstance) sanityCheckPVClusterData(pvList []corev1.PersistentVolume)
 
 type ObjectStorePVDownloader struct{}
 
-func (s ObjectStorePVDownloader) DownloadPVs(ctx context.Context, r client.Reader,
+func (s ObjectStorePVDownloader) DownloadPVsAndPVCs(ctx context.Context, r client.Reader,
 	objStoreGetter ObjectStoreGetter, s3Profile, s3KeyPrefix string,
-	debugTag string) ([]corev1.PersistentVolume, error) {
+	debugTag string) ([]corev1.PersistentVolume, []corev1.PersistentVolumeClaim, error) {
 	objectStore, err := objStoreGetter.ObjectStore(ctx, r, s3Profile, debugTag)
 	if err != nil {
-		return nil, fmt.Errorf("error when downloading PVs, err %w", err)
+		return nil, nil, fmt.Errorf("error when downloading PVs, err %w", err)
 	}
 
-	return objectStore.DownloadPVs(s3KeyPrefix)
+	return objectStore.DownloadPVsAndPVCs(s3KeyPrefix)
+}
+
+func (v *VRGInstance) restorePVAndPVCClusterData(pvList []corev1.PersistentVolume,
+	pvcList []corev1.PersistentVolumeClaim) error {
+	err := v.restorePVClusterData(pvList)
```

Comment:
Should we be super sensitive if the PV and PVC counts differ (more importantly if the PVC count is higher than the PV count) and error out before we restore?

There is no way out for the user to rectify the situation than going into the S3 store and updating the objects, so would be an unrecoverable error. Just thinking aloud at the moment.

Looking at the backup, the only possibility would be PV count greater than PVC count, as we upload PV first (which is correct). Anyway, leaving this thought here...

Reply:
We probably need a way to validate that the data is valid in the s3Store before the user attempts a failover.  But I do agree that we should have at least a check here.